### PR TITLE
refactor(blockifier): use raw contract class in class info

### PR DESCRIPTION
**Pull request by noaov1**

### CODE REVIEW

**Comment by ArniStarkware** [comment]
*[`crates/starknet_api/src/contract_class.rs` line 25 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1730#-OAlKsqh2XYnPcAcKDPD:-OAlKsqh2XYnPcAcKDPE:b-eurkta) ([raw file](https://github.com/starkware-libs/sequencer/blob/e9f3ebe7f94c43b39f41db655c8fef1a315bf9f7/crates/starknet_api/src/contract_class.rs#L25)):*
> ```XML
> 
> /// Represents a raw Starknet contract class.
> #[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize, derive_more::From)]
> ```

What does this do?



_Code quote:_
```XML
 derive_more::From
```

**Review by ArniStarkware — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1730)* status: 0 of 16 files reviewed, 4 unresolved discussions (waiting on @dorimedini-starkware and @noaov1)

___
*[`crates/blockifier/src/execution/contract_class.rs` line 305 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1730#-OAlP9Rc2o4Va-NFvJTJ:-OAlP9Rc2o4Va-NFvJTK:bfxm5lz) ([raw file](https://github.com/starkware-libs/sequencer/blob/6797cae0aa1c2e46fea7b4430025239228860b61/crates/blockifier/src/execution/contract_class.rs#L305)):*
> ```XML
>             bytecode_segment_lengths: NestedIntList::Leaf(0),
>         }))
>     }
> ```

Can be it's own PR - right? 



_Code quote:_
```XML
    /// Returns an empty contract class for testing purposes.
    #[cfg(any(feature = "testing", test))]
    pub fn empty_for_testing() -> Self {
        Self(Arc::new(ContractClassV1Inner {
            program: Default::default(),
            entry_points_by_type: Default::default(),
            hints: Default::default(),
            compiler_version: Default::default(),
            bytecode_segment_lengths: NestedIntList::Leaf(0),
        }))
    }
```
___
*[`crates/blockifier/src/test_utils/contracts.rs` line 166 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1730#-OAlNtFu3wIrBG9zol4A:-OAlNtFu3wIrBG9zol4B:b-p2bhxa) ([raw file](https://github.com/starkware-libs/sequencer/blob/e9f3ebe7f94c43b39f41db655c8fef1a315bf9f7/crates/blockifier/src/test_utils/contracts.rs#L166)):*
> ```XML
>     pub fn get_runnable_class(&self) -> RunnableContractClass {
>         self.get_class().try_into().unwrap()
>     }
> ```

Before this change, `get_runnable_class` was just a `load`.
Now, it is a conversion - i.e., it requires actual computations.

For a test util, it would be worth it to do it directly.,



_Code quote:_
```XML
    pub fn get_runnable_class(&self) -> RunnableContractClass {
        self.get_class().try_into().unwrap()
    }
```
___
*[`crates/blockifier/src/transaction/account_transactions_test.rs` line 751 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1730#-OAlLagUEtshS_WJY8oQ:-OAlLagUEtshS_WJY8oR:b2idc0j) ([raw file](https://github.com/starkware-libs/sequencer/blob/e9f3ebe7f94c43b39f41db655c8fef1a315bf9f7/crates/blockifier/src/transaction/account_transactions_test.rs#L751)):*
> ```XML
>         create_test_init_data(chain_info, CairoVersion::Cairo0);
>     let class_hash = class_hash!(0xdeadeadeaf72_u128);
>     let contract_class = FeatureContract::FaultyAccount(CairoVersion::Cairo1).get_class();
> ```

Does this change the functionality?



_Code quote:_
```XML
    let contract_class = FeatureContract::FaultyAccount(CairoVersion::Cairo1).get_class();
```
___
*[`crates/blockifier_reexecution/src/state_reader/compile.rs` line 95 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1730#-OAlLH0D21Dn2V-KQ78o:-OAlLH0EAlFforXZs972:b-7n0m76) ([raw file](https://github.com/starkware-libs/sequencer/blob/e9f3ebe7f94c43b39f41db655c8fef1a315bf9f7/crates/blockifier_reexecution/src/state_reader/compile.rs#L95)):*
> ```XML
>             usize::MAX,
>         )
>         // TODO(Aviv): Reconsider the unwrap.
> ```

@AvivYossef-starkware / @avivg-starkware 



_Code quote:_
```XML
// TODO(Aviv): Reconsider the unwrap.
```

**Comment by ArniStarkware** [comment]
*[`crates/blockifier/src/execution/contract_class.rs` line 524 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1730#-OAlQLrc4ofBNM_83u1V:-OAlQLrc4ofBNM_83u1W:b-1rfja2) ([raw file](https://github.com/starkware-libs/sequencer/blob/e9f3ebe7f94c43b39f41db655c8fef1a315bf9f7/crates/blockifier/src/execution/contract_class.rs#L524)):*
> ```XML
>     sierra_program_length: usize,
>     abi_length: usize,
> }
> ```

I think this Todo is finally done here!

@ayeletstarkware - be aware.



_Suggestion:_
```XML
#[derive(Clone, Debug)]
pub struct ClassInfo {
    contract_class: ContractClass,
    sierra_program_length: usize,
    abi_length: usize,
}
```

**Review by noaov1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1730)* status: 0 of 16 files reviewed, 1 unresolved discussion (waiting on @ArniStarkware and @dorimedini-starkware)

*[a discussion](https://reviewable.io/reviews/starkware-libs/sequencer/1730#-OAkzJ0b-mh3-n5u1SMs:-OAkzJ0b-mh3-n5u1SMt:biytyfq) (no related file):*
@AvivYossef-starkware @aner-starkware FYI

___

**Review by dorimedini-starkware — CHANGES_REQUESTED** [review]
> Reviewed 13 of 14 files at r1, 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1730)* status: all files reviewed, 10 unresolved discussions (waiting on @ArniStarkware, @dorimedini-starkware, and @noaov1)

___
*[`crates/blockifier/src/execution/contract_class.rs` line 548 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1730#-OAlOfgnEMHGk1_IiPhP:-OAlOfgnEMHGk1_IiPhQ:b-776mcs) ([raw file](https://github.com/starkware-libs/sequencer/blob/e9f3ebe7f94c43b39f41db655c8fef1a315bf9f7/crates/blockifier/src/execution/contract_class.rs#L548)):*
> ```XML
>                 .as_array()
>                 .map(|arr| arr.len())
>                 .expect("The program data must be an array."),
> ```

can this logic be part of the V0 class?



_Code quote:_
```XML
            ContractClass::V0(contract_class) => contract_class
                .program
                .data
                .as_array()
                .map(|arr| arr.len())
                .expect("The program data must be an array."),
```
___
*[`crates/blockifier/src/test_utils/contracts.rs` line 13 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1730#-OAlOnas-IlMOG1o6-d-:-OAlOnas-IlMOG1o6-d0:bflycw3) ([raw file](https://github.com/starkware-libs/sequencer/blob/e9f3ebe7f94c43b39f41db655c8fef1a315bf9f7/crates/blockifier/src/test_utils/contracts.rs#L13)):*
> ```XML
> use strum_macros::EnumIter;
> 
> use super::struct_impls::LoadFile;
> ```

in test modules it's ok to use `super` I think but please be consistent
you have `crate::test_utils::...` imports here below



_Code quote:_
```XML
super::struct_impls
```
___
*[`crates/blockifier/src/transaction/account_transactions_test.rs` line 764 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1730#-OAlP4Vs8mfZhasT8f7l:-OAlP4Vs8mfZhasT8f7m:b-5ab039) ([raw file](https://github.com/starkware-libs/sequencer/blob/e9f3ebe7f94c43b39f41db655c8fef1a315bf9f7/crates/blockifier/src/transaction/account_transactions_test.rs#L764)):*
> ```XML
>         .set_contract_class(
>             class_hash,
>             TryInto::<RunnableContractClass>::try_into(contract_class.clone()).unwrap(),
> ```

if this doesn't work, why not?



_Suggestion:_
```XML
contract_class.clone().try_into().unwrap(),
```
___
*[`crates/blockifier_reexecution/src/state_reader/compile.rs` line 105 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1730#-OAlPyyRCsFqex9Ub_jH:-OAlPyyRCsFqex9Ub_jI:bh87lkn) ([raw file](https://github.com/starkware-libs/sequencer/blob/e9f3ebe7f94c43b39f41db655c8fef1a315bf9f7/crates/blockifier_reexecution/src/state_reader/compile.rs#L105)):*
> ```XML
> ) -> StateResult<ContractClass> {
>     let as_str = decode_reader(legacy.program).unwrap();
>     let program: Program = serde_json::from_str(&as_str).unwrap();
> ```

what is this change? why was there a conversion to `bytes` previously?



_Code quote:_
```XML
 serde_json::from_str(&as_str).unwrap();
```
___
*[`crates/papyrus_execution/src/lib.rs` line 799 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1730#-OAlQUBU7ATZuHzhKmxl:-OAlQUBU7ATZuHzhKmxm:bnd8rei) ([raw file](https://github.com/starkware-libs/sequencer/blob/e9f3ebe7f94c43b39f41db655c8fef1a315bf9f7/crates/papyrus_execution/src/lib.rs#L799)):*
> ```XML
>                 DEPRECATED_CONTRACT_SIERRA_SIZE,
>                 abi_length,
>             )
> ```

can `deprecated_class.into()` panic? why is there no more room for error?
+ please add reviewer from papyrus team?



_Code quote:_
```XML
            let class_info = ClassInfo::new(
                &deprecated_class.into(),
                DEPRECATED_CONTRACT_SIERRA_SIZE,
                abi_length,
            )
```
___
*[`crates/starknet_api/src/contract_class.rs` line 25 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1730#-OAlKsqh2XYnPcAcKDPD:-OAlPTa26rQ-ew6VexM9:b-715522) ([raw file](https://github.com/starkware-libs/sequencer/blob/e9f3ebe7f94c43b39f41db655c8fef1a315bf9f7/crates/starknet_api/src/contract_class.rs#L25)):*
<details><summary><i>Previously, ArniStarkware (Arnon Hod) wrote…</i></summary><blockquote>

What does this do?
</blockquote></details>

I think it's more readable to explicitly implement `From<RunnableContractClass> for ContractClass`
that way you maybe won't need the `TryInfo::<RunnableContractClass>::...` turbofish above?

**Review by ArniStarkware — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1730)* status: all files reviewed, 11 unresolved discussions (waiting on @dorimedini-starkware and @noaov1)

___
*[`crates/blockifier/src/execution/contract_class.rs` line 538 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1730#-OAlR9jI2sESlyM3I2O5:-OAlR9jJ8WPpeCtn4eG6:b-w1pwnj) ([raw file](https://github.com/starkware-libs/sequencer/blob/e9f3ebe7f94c43b39f41db655c8fef1a315bf9f7/crates/blockifier/src/execution/contract_class.rs#L538)):*
> ```XML
>         Ok(Self { contract_class: contract_class.clone(), sierra_program_length, abi_length })
>     }
> }
> ```

These structs became precisely the same after this PR (nice!) - except for implementations.

I see some of the impls can be moved to snapi. Is it true for all of the implementations?

TL;DR - delete the struct from blockifier and use the struct from snapi instead.



_Code quote:_
```XML
impl TryFrom<starknet_api::contract_class::ClassInfo> for ClassInfo {
    type Error = ProgramError;

    fn try_from(class_info: starknet_api::contract_class::ClassInfo) -> Result<Self, Self::Error> {
        let starknet_api::contract_class::ClassInfo {
            contract_class,
            sierra_program_length,
            abi_length,
        } = class_info;

        Ok(Self { contract_class: contract_class.clone(), sierra_program_length, abi_length })
    }
}
```

**Review by ArniStarkware — CHANGES_REQUESTED** [review]
> Reviewed 4 of 14 files at r1, 1 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1730)* status: all files reviewed, 13 unresolved discussions (waiting on @dorimedini-starkware and @noaov1)

___
*[`crates/blockifier/src/transaction/transactions.rs` line 156 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1730#-OAlTSym8lOa5mxJtxT2:-OAlTSym8lOa5mxJtxT3:b-l31rnx) ([raw file](https://github.com/starkware-libs/sequencer/blob/e9f3ebe7f94c43b39f41db655c8fef1a315bf9f7/crates/blockifier/src/transaction/transactions.rs#L156)):*
> ```XML
>         let declare_version = declare_tx.version();
>         // Verify contract class version.
>         if !is_cairo1(&class_info.contract_class().try_into()?) {
> ```

We do not need to do a conversion into runnable contract class just to check the cairo version.



_Code quote:_
```XML
!is_cairo1(&class_info.contract_class().try_into()?)
```
___
*[`crates/blockifier/src/transaction/transactions_test.rs` line 1569 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1730#-OAlSuCg2j8YopRwWMx7:-OAlSuCg2j8YopRwWMx8:beuw6sy) ([raw file](https://github.com/starkware-libs/sequencer/blob/e9f3ebe7f94c43b39f41db655c8fef1a315bf9f7/crates/blockifier/src/transaction/transactions_test.rs#L1569)):*
> ```XML
>     // Verify class declaration.
>     let contract_class_from_state = state.get_compiled_contract_class(class_hash).unwrap();
>     assert_eq!(contract_class_from_state, class_info.contract_class().try_into().unwrap());
> ```

There is computation done here. Can It be loaded more directly from somewhere?



_Code quote:_
```XML
class_info.contract_class().try_into().unwrap()
```

**Comment by ArniStarkware** [comment]
*[`crates/blockifier/src/transaction/account_transactions_test.rs` line 751 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1730#-OAlLagUEtshS_WJY8oQ:-OAlTtBXBboWiLut5L0k:b-azfm8z) ([raw file](https://github.com/starkware-libs/sequencer/blob/e9f3ebe7f94c43b39f41db655c8fef1a315bf9f7/crates/blockifier/src/transaction/account_transactions_test.rs#L751)):*
<details><summary><i>Previously, ArniStarkware (Arnon Hod) wrote…</i></summary><blockquote>

Does this change the functionality?
</blockquote></details>

Document the fact that the contract used here does not matter.

Also - isn't there an empty feature contract?

**Review by ArniStarkware — COMMENTED** [review]
> Reviewed 2 of 14 files at r1.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1730)* status: all files reviewed, 13 unresolved discussions (waiting on @dorimedini-starkware and @noaov1)

**Review by noaov1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1730)* status: all files reviewed, 13 unresolved discussions (waiting on @ArniStarkware, @ayeletstarkware, and @dorimedini-starkware)

___
*[`crates/blockifier/src/execution/contract_class.rs` line 305 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1730#-OAlP9Rc2o4Va-NFvJTJ:-OAlWrN-EYXEfDlNcB18:b-aa42pd) ([raw file](https://github.com/starkware-libs/sequencer/blob/6797cae0aa1c2e46fea7b4430025239228860b61/crates/blockifier/src/execution/contract_class.rs#L305)):*
<details><summary><i>Previously, ArniStarkware (Arnon Hod) wrote…</i></summary><blockquote>

Can be it's own PR - right?
</blockquote></details>

Yes. This PR made it an unused method. I'll add it in a separate PR. 
___
*[`crates/blockifier/src/execution/contract_class.rs` line 524 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1730#-OAlQLrc4ofBNM_83u1V:-OAlWp-_5bMFIv2IqjKu:b-cott9w) ([raw file](https://github.com/starkware-libs/sequencer/blob/e9f3ebe7f94c43b39f41db655c8fef1a315bf9f7/crates/blockifier/src/execution/contract_class.rs#L524)):*
<details><summary><i>Previously, ArniStarkware (Arnon Hod) wrote…</i></summary><blockquote>

I think this Todo is finally done here!

@ayeletstarkware - be aware.
</blockquote></details>

Why? :thinking: 
___
*[`crates/blockifier/src/execution/contract_class.rs` line 538 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1730#-OAlR9jI2sESlyM3I2O5:-OAlWf7vDlptXciRb9Ux:b1n0dvd) ([raw file](https://github.com/starkware-libs/sequencer/blob/e9f3ebe7f94c43b39f41db655c8fef1a315bf9f7/crates/blockifier/src/execution/contract_class.rs#L538)):*
Yes that's the plan in the upcoming PRs.

> I see some of the impls can be moved to snapi. Is it true for all of the implementations?

It should be the case. I tried to avoid using vm dependency and blockifier code. 
___
*[`crates/blockifier/src/execution/contract_class.rs` line 548 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1730#-OAlOfgnEMHGk1_IiPhP:-OAlWn3W5ealoabA1Rw4:b-896fix) ([raw file](https://github.com/starkware-libs/sequencer/blob/e9f3ebe7f94c43b39f41db655c8fef1a315bf9f7/crates/blockifier/src/execution/contract_class.rs#L548)):*
<details><summary><i>Previously, dorimedini-starkware wrote…</i></summary><blockquote>

can this logic be part of the V0 class?
</blockquote></details>

Done.
___
*[`crates/blockifier/src/test_utils/contracts.rs` line 13 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1730#-OAlOnas-IlMOG1o6-d-:-OAlYW4G6-NXqHZFa5ks:b-3jq4rr) ([raw file](https://github.com/starkware-libs/sequencer/blob/e9f3ebe7f94c43b39f41db655c8fef1a315bf9f7/crates/blockifier/src/test_utils/contracts.rs#L13)):*
<details><summary><i>Previously, dorimedini-starkware wrote…</i></summary><blockquote>

in test modules it's ok to use `super` I think but please be consistent
you have `crate::test_utils::...` imports here below
</blockquote></details>

Nope. Fixed. 
___
*[`crates/blockifier/src/test_utils/contracts.rs` line 166 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1730#-OAlNtFu3wIrBG9zol4A:-OAlYsms43c47UQ-KfDz:bskkzwj) ([raw file](https://github.com/starkware-libs/sequencer/blob/e9f3ebe7f94c43b39f41db655c8fef1a315bf9f7/crates/blockifier/src/test_utils/contracts.rs#L166)):*
<details><summary><i>Previously, ArniStarkware (Arnon Hod) wrote…</i></summary><blockquote>

Before this change, `get_runnable_class` was just a `load`.
Now, it is a conversion - i.e., it requires actual computations.

For a test util, it would be worth it to do it directly.,
</blockquote></details>

It is the same behaviour. Previously, we loaded the json file and converted.  For cairo 0 we used it implicitly with [deserialize_program](https://github.com/starkware-libs/sequencer/blob/8823d312e735077259a677be3f301930a4e95e66/crates/blockifier/src/execution/contract_class.rs#L211) and for cairo 1 we use it explicitly [here](https://github.com/starkware-libs/sequencer/blob/8823d312e735077259a677be3f301930a4e95e66/crates/blockifier/src/execution/contract_class.rs#L290).
___
*[`crates/blockifier/src/transaction/account_transactions_test.rs` line 751 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1730#-OAlLagUEtshS_WJY8oQ:-OAlZ58m0mVSOW6GT__0:b5pgfh3) ([raw file](https://github.com/starkware-libs/sequencer/blob/e9f3ebe7f94c43b39f41db655c8fef1a315bf9f7/crates/blockifier/src/transaction/account_transactions_test.rs#L751)):*
<details><summary><i>Previously, ArniStarkware (Arnon Hod) wrote…</i></summary><blockquote>

Document the fact that the contract used here does not matter.

Also - isn't there an empty feature contract?
</blockquote></details>

Changed. Thanks!
___
*[`crates/blockifier/src/transaction/account_transactions_test.rs` line 764 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1730#-OAlP4Vs8mfZhasT8f7l:-OAlfWy-20O4JydMs7TN:bw0iqnw) ([raw file](https://github.com/starkware-libs/sequencer/blob/e9f3ebe7f94c43b39f41db655c8fef1a315bf9f7/crates/blockifier/src/transaction/account_transactions_test.rs#L764)):*
<details><summary><i>Previously, dorimedini-starkware wrote…</i></summary><blockquote>

if this doesn't work, why not?
</blockquote></details>

For some mysterious reason, now it's working
___
*[`crates/blockifier/src/transaction/transactions.rs` line 156 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1730#-OAlTSym8lOa5mxJtxT2:-OAldKDR5zw9YAgBLTTD:b-z4vojd) ([raw file](https://github.com/starkware-libs/sequencer/blob/e9f3ebe7f94c43b39f41db655c8fef1a315bf9f7/crates/blockifier/src/transaction/transactions.rs#L156)):*
<details><summary><i>Previously, ArniStarkware (Arnon Hod) wrote…</i></summary><blockquote>

We do not need to do a conversion into runnable contract class just to check the cairo version.
</blockquote></details>

Good point. 
Will be handled in a separate PR (added a TODO). 
(I'll add a trait for the contract classes)
___
*[`crates/blockifier/src/transaction/transactions_test.rs` line 1569 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1730#-OAlSuCg2j8YopRwWMx7:-OAlfRAmBVXxUHc7tuRP:b-odjpyn) ([raw file](https://github.com/starkware-libs/sequencer/blob/e9f3ebe7f94c43b39f41db655c8fef1a315bf9f7/crates/blockifier/src/transaction/transactions_test.rs#L1569)):*
<details><summary><i>Previously, ArniStarkware (Arnon Hod) wrote…</i></summary><blockquote>

There is computation done here. Can It be loaded more directly from somewhere?
</blockquote></details>

I am not sure I understand. WDYM?
___
*[`crates/blockifier_reexecution/src/state_reader/compile.rs` line 105 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1730#-OAlPyyRCsFqex9Ub_jH:-OAle7Fc9zXvP5YsB6np:b-ybxeft) ([raw file](https://github.com/starkware-libs/sequencer/blob/e9f3ebe7f94c43b39f41db655c8fef1a315bf9f7/crates/blockifier_reexecution/src/state_reader/compile.rs#L105)):*
<details><summary><i>Previously, dorimedini-starkware wrote…</i></summary><blockquote>

what is this change? why was there a conversion to `bytes` previously?
</blockquote></details>

That is the VM program API. We now use the starknet api [program](https://github.com/starkware-libs/sequencer/blob/cfbb00e092529af837da97966ed1e9991dac5ffd/crates/starknet_api/src/deprecated_contract_class.rs#L127).
___
*[`crates/papyrus_execution/src/lib.rs` line 799 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1730#-OAlQUBU7ATZuHzhKmxl:-OAleKxqAPS-opY4GLbW:b-anttfu) ([raw file](https://github.com/starkware-libs/sequencer/blob/e9f3ebe7f94c43b39f41db655c8fef1a315bf9f7/crates/papyrus_execution/src/lib.rs#L799)):*
We don't need to convert the sn api program to the vm program anymore. 

> can `deprecated_class.into()` panic?

No. It converts `DeprecatedContractClass` to it's enum parent `ContractClass`. 
___
*[`crates/starknet_api/src/contract_class.rs` line 25 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1730#-OAlKsqh2XYnPcAcKDPD:-OAledp46jjv3ZMdL5AT:blv6myw) ([raw file](https://github.com/starkware-libs/sequencer/blob/e9f3ebe7f94c43b39f41db655c8fef1a315bf9f7/crates/starknet_api/src/contract_class.rs#L25)):*
<details><summary><i>Previously, dorimedini-starkware wrote…</i></summary><blockquote>

I think it's more readable to explicitly implement `From<RunnableContractClass> for ContractClass`
that way you maybe won't need the `TryInfo::<RunnableContractClass>::...` turbofish above?
</blockquote></details>

It allows us to convert `DeprecatedContractClass` or `CasmContractClass` to its parent enum `ContractClass`.

**Review by ArniStarkware — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1730)* status: 9 of 17 files reviewed, 8 unresolved discussions (waiting on @ayeletstarkware and @dorimedini-starkware)

___
*[`crates/blockifier/src/execution/contract_class.rs` line 524 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1730#-OAlQLrc4ofBNM_83u1V:-OAli63dC7gGt-OjgZzd:b19zx1f) ([raw file](https://github.com/starkware-libs/sequencer/blob/e9f3ebe7f94c43b39f41db655c8fef1a315bf9f7/crates/blockifier/src/execution/contract_class.rs#L524)):*
<details><summary><i>Previously, noaov1 (Noa Oved) wrote…</i></summary><blockquote>

Why? :thinking:
</blockquote></details>

I understand this todo as:
Use a `ContractClass` class that is just raw data. This is exactly what this PR does (It replaces the `ContractClass` with felts [not bytes - but still] instead of CairoVm stuff).

If this is indeed the intention - this TODO is completed. 
___
*[`crates/blockifier/src/transaction/transactions_test.rs` line 1569 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1730#-OAlSuCg2j8YopRwWMx7:-OAljTJ2CnqZxx4S9NSX:bm5s1hr) ([raw file](https://github.com/starkware-libs/sequencer/blob/e9f3ebe7f94c43b39f41db655c8fef1a315bf9f7/crates/blockifier/src/transaction/transactions_test.rs#L1569)):*
<details><summary><i>Previously, noaov1 (Noa Oved) wrote…</i></summary><blockquote>

I am not sure I understand. WDYM?
</blockquote></details>

NVM. I tried to suggest some optimization.
I think there is no clean way to do this optimization - and it is not worth thinking about.

**Review by dorimedini-starkware — CHANGES_REQUESTED** [review]
> Reviewed 8 of 8 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1730)* status: all files reviewed, 4 unresolved discussions (waiting on @ArniStarkware, @ayeletstarkware, and @noaov1)

___
*[`crates/starknet_api/src/deprecated_contract_class.rs` line 35 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/1730#-OAlk8DrE-oEx6o-YQ7l:-OAlk8DrE-oEx6o-YQ7m:bxoxnoa) ([raw file](https://github.com/starkware-libs/sequencer/blob/8253a718656a85abf49b96320054e14b59594e23/crates/starknet_api/src/deprecated_contract_class.rs#L35)):*
> ```XML
>             .as_array()
>             .map(|arr| arr.len())
>             .expect("The program data must be an array.")
> ```

you are `expect`ing anyway, no point in `map` first



_Suggestion:_
```XML
            .data
            .as_array()
            .expect("The program data must be an array.")
            .len()
            
```

**Review by ArniStarkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1730)* status: all files reviewed, 3 unresolved discussions (waiting on @ayeletstarkware and @noaov1)

___
*[`crates/starknet_api/src/contract_class.rs` line 25 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1730#-OAlKsqh2XYnPcAcKDPD:-OAlkQ6LBBL3TfnekHqD:bewky5k) ([raw file](https://github.com/starkware-libs/sequencer/blob/e9f3ebe7f94c43b39f41db655c8fef1a315bf9f7/crates/starknet_api/src/contract_class.rs#L25)):*
<details><summary><i>Previously, noaov1 (Noa Oved) wrote…</i></summary><blockquote>

It allows us to convert `DeprecatedContractClass` or `CasmContractClass` to its parent enum `ContractClass`.
</blockquote></details>

Seems like the issue @dorimedini-starkware is talking about is different.

**Review by dorimedini-starkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1730)* status: all files reviewed, 3 unresolved discussions (waiting on @ayeletstarkware and @noaov1)

___
*[`crates/starknet_api/src/contract_class.rs` line 25 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1730#-OAlKsqh2XYnPcAcKDPD:-OAllYpDCP18T0NkZzHc:buwpc7z) ([raw file](https://github.com/starkware-libs/sequencer/blob/e9f3ebe7f94c43b39f41db655c8fef1a315bf9f7/crates/starknet_api/src/contract_class.rs#L25)):*
<details><summary><i>Previously, ArniStarkware (Arnon Hod) wrote…</i></summary><blockquote>

Seems like the issue @dorimedini-starkware is talking about is different.
</blockquote></details>

yeah my mistake, I didn't know what this did :)
the turbofish wasn't needed anyway

**Review by noaov1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1730)* status: all files reviewed, 3 unresolved discussions (waiting on @ayeletstarkware and @dorimedini-starkware)

___
*[`crates/blockifier/src/execution/contract_class.rs` line 524 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1730#-OAlQLrc4ofBNM_83u1V:-OAlmK_89jdOoVmnJUgg:b-pfpp8o) ([raw file](https://github.com/starkware-libs/sequencer/blob/e9f3ebe7f94c43b39f41db655c8fef1a315bf9f7/crates/blockifier/src/execution/contract_class.rs#L524)):*
<details><summary><i>Previously, ArniStarkware (Arnon Hod) wrote…</i></summary><blockquote>

I understand this todo as:
Use a `ContractClass` class that is just raw data. This is exactly what this PR does (It replaces the `ContractClass` with felts [not bytes - but still] instead of CairoVm stuff).

If this is indeed the intention - this TODO is completed.
</blockquote></details>

Ayelet does not remember. IIUC, the intention was to store it as bytes. I will leave it here for now. 
___
*[`crates/starknet_api/src/deprecated_contract_class.rs` line 35 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/1730#-OAlk8DrE-oEx6o-YQ7l:-OAllzaD8Pu971N7ygCl:b-896fix) ([raw file](https://github.com/starkware-libs/sequencer/blob/8253a718656a85abf49b96320054e14b59594e23/crates/starknet_api/src/deprecated_contract_class.rs#L35)):*
<details><summary><i>Previously, dorimedini-starkware wrote…</i></summary><blockquote>

you are `expect`ing anyway, no point in `map` first
</blockquote></details>

Done.

**Review by ArniStarkware — APPROVED** [review]
:lgtm:

> Reviewed 2 of 14 files at r1, 1 of 3 files at r2, 7 of 8 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1730#-:-OAm6z5i33iMxP_5O5rv:bnfp4nl)* status: all files reviewed, 2 unresolved discussions (waiting on @ayeletstarkware and @dorimedini-starkware)

**Review by ArniStarkware — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1730)* status: all files reviewed, 2 unresolved discussions (waiting on @ayeletstarkware and @dorimedini-starkware)

___
*[`crates/blockifier/src/execution/contract_class.rs` line 551 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/1730#-OAm7SmZDvru6FZRQ0tq:-OAm7SmZDvru6FZRQ0tr:bb4biu9) ([raw file](https://github.com/starkware-libs/sequencer/blob/e145ec539e170aac6fd460feec1111d8c191f723/crates/blockifier/src/execution/contract_class.rs#L551)):*
> ```XML
>             ContractClass::V0(contract_class) => contract_class.bytecode_length(),
>             ContractClass::V1(contract_class) => contract_class.bytecode.len(),
>         }
> ```

This logic can be a part of `ContractClass`.
Can wait to the PR where you squash the blockifier's `ClassInfo` into the snapi `ClassInfo`). 



_Code quote:_
```XML
        match &self.contract_class {
            ContractClass::V0(contract_class) => contract_class.bytecode_length(),
            ContractClass::V1(contract_class) => contract_class.bytecode.len(),
        }
```

**Review by dorimedini-starkware — APPROVED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1730)* status: all files reviewed, 1 unresolved discussion (waiting on @ayeletstarkware)
